### PR TITLE
conv3d_transpose in tf, th, and cntk

### PR DESCRIPTION
**Pull request by ncullen93**
Implemented Conv3DTranspose (w/ Deconv3D and Deconvolution3D aliases) in tensorflow, theano, and cntk backends, along with the appropriate abstract layer. This is implemented in the Keras 2.0 API and you'll notice how similar these operations are to the relevant 2D ones.

### CODE REVIEW

**Comment by gewoonrik** [comment]
could you please add some tests?

File: tests/keras/layers/convolutional_test.py [hunk 1]
```diff
@@ -389,6 +389,51 @@ def test_convolution_3d():
...
+            if padding == 'same' and strides != (1, 1, 1):
+                continue
+            layer_test(convolutional.Deconvolution3D,
+                       kwargs={'filters': filters,
+                               'kernel_size': 3,
+                               'padding': padding,
+                               'strides': strides,
+                               'data_format': 'channels_last'},
+                       input_shape=(num_samples, num_row, num_col, num_depth, stack_size),
+                       fixed_batch_size=True)
```
Comment (fchollet) [inline tests/keras/layers/convolutional_test.py]: Should not require a fixed batch size

File: tests/keras/layers/convolutional_test.py [hunk 2]
```diff
@@ -389,6 +389,51 @@ def test_convolution_3d():
...
+    stack_size = 3
+    num_row = 5
+    num_col = 6
+    num_depth = 7
+
+    for padding in _convolution_paddings:
+        for strides in [(1, 1, 1), (2, 2, 2)]:
+            if padding == 'same' and strides != (1, 1, 1):
+                continue
+            layer_test(convolutional.Deconvolution3D,
```
Comment (fchollet) [inline tests/keras/layers/convolutional_test.py]: Use `Conv3DTranspose`

File: tests/keras/layers/convolutional_test.py [hunk 3]
```diff
@@ -389,6 +389,51 @@ def test_convolution_3d():
...
+                       kwargs={'filters': filters,
+                               'kernel_size': 3,
+                               'padding': padding,
+                               'strides': strides,
+                               'data_format': 'channels_last'},
```
Comment (fchollet) [inline tests/keras/layers/convolutional_test.py]: Please add a test case for `channels_first`

File: keras/layers/convolutional.py [hunk 4]
```diff
@@ -806,6 +806,226 @@ def get_config(self):
...
+    of a normal convolution, i.e., from something that has the shape of the
+    output of some convolution to something that has the shape of its input
+    while maintaining a connectivity pattern that is compatible with
+    said convolution.
+
+    When using this layer as the first layer in a model,
+    provide the keyword argument `input_shape`
+    (tuple of integers, does not include the sample axis),
+    e.g. `input_shape=(128, 128, 128, 3)` if `data_format="channels_last"`
+    or (3, 128, 128, 128) if `data_format="channels_first"`.
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Specify that this would be for 128x128x128 volumes with 3 channels

File: keras/layers/convolutional.py [hunk 5]
```diff
@@ -806,6 +806,226 @@ def get_config(self):
...
+            Can be a single integer to specify the same value for
+            all spatial dimensions.
+            Specifying any stride value != 1 is incompatible with specifying
+            any `dilation_rate` value != 1.
+        padding: one of `"valid"` or `"same"` (case-insensitive).
+        data_format: A string,
+            one of `channels_last` (default) or `channels_first`.
+            The ordering of the dimensions in the inputs.
+            `channels_last` corresponds to inputs with shape
+            `(batch, height, width, channels)` while `channels_first`
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Shapes mentioned in the docstring are generally 2D; should be 3D

File: keras/layers/convolutional.py [hunk 6]
```diff
@@ -806,6 +806,226 @@ def get_config(self):
...
+            the output of the layer (its "activation").
+            (see [regularizer](../regularizers.md)).
+        kernel_constraint: Constraint function applied to the kernel matrix
+            (see [constraints](../constraints.md)).
+        bias_constraint: Constraint function applied to the bias vector
+            (see [constraints](../constraints.md)).
+
+    # Input shape
+        5D tensor with shape:
+        `(batch, channels, rows, cols, depth)` if data_format='channels_first'
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Depth should come before rows and cols (this might need to be fixed elsewhere as well)

File: keras/layers/convolutional.py [hunk 7]
```diff
@@ -806,6 +806,226 @@ def get_config(self):
...
+            self.bias = None
+        # Set input spec.
+        self.input_spec = InputSpec(ndim=5, axes={channel_axis: input_dim})
+        self.built = True
+
+    def call(self, inputs):
+        input_shape = K.shape(inputs)
+        batch_size = input_shape[0]
+        if self.data_format == 'channels_first':
+            h_axis, w_axis, d_axis = 2, 3, 4
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Incorrect axis naming; use https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/convolutional.py#L1386 as a reference

**Comment by ncullen93** [comment]
Regarding tests - I added more conditions as requested, but I originally just copied the `test_conv2d_transpose` function verbatim :)

File: tests/keras/layers/convolutional_test.py [hunk 8]
```diff
@@ -389,6 +389,52 @@ def test_convolution_3d():
...
+def test_conv3d_transpose():
+    filters = 2
+    stack_size = 3
+    num_row = 5
+    num_col = 6
+    num_depth = 7
+
+    for padding in _convolution_paddings:
+        for strides in [(1, 1, 1), (2, 2, 2)]:
+            for kernel_size in [2,3,4]:
```
Comment (fchollet) [inline tests/keras/layers/convolutional_test.py]: I don't think that's necessary, a single kernel size (3) should be ok. Such a nested loop is very expensive

**Comment by jocicmarko** [comment]
Well documented and neatly done PR. LGTM
